File: native/scheduler.cpp
#include "scheduler.h"
#include <chrono>
#include "execute.h"
#include <iostream>
#include <vector>
#include <thread>
#include <linux/perf_event.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <cstring>
#include <asm/unistd.h>


int set_cpu_mask(pid_t pid, int from, int end) {
    return 1;
    cpu_set_t cpuset;
    CPU_ZERO(&cpuset);  // 初始化 CPU 集合

    for (int i = from; i <= end; i++) {
        CPU_SET(i, &cpuset);  // 将大集群中的核心添加到集合中
    }

    if (sched_setaffinity(pid, sizeof(cpuset), &cpuset) == -1) {
        perror("sched_setaffinity");
        return 1;
    }
    return 0;
}

int set_cpu_mask_big(pid_t pid) {
    return set_cpu_mask(pid, 0, 3);
}

int set_cpu_mask_little(pid_t pid) {
    return set_cpu_mask(pid, 4, 7);
}

int set_cpu_mask_all(pid_t pid) {
    return set_cpu_mask(pid, 0, 7);
}

std::vector<int> argsort(std::vector<double> input) {
    int h = input.size();
    std::vector<int> index(h);

    for (int i = 0; i < h; ++i)
        index[i] = i;

    for (int i = h - 1; i > 0; --i) {
        int min = 0;
        for (int j = 1; j <= i; ++j) {
            if (input[index[j]] < input[index[min]])
                min = j;
        }

        int temp = index[i];
        index[i] = index[min];
        index[min] = temp;
    }
    return index;
}

std::vector<std::string> split(const std::string& input, char delimiter) {  // 该split对单个分隔符进行分隔
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(input);

    while (std::getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }

    return tokens;
}

std::vector<std::string> split(const std::string& str) {  // 该split对一个或多个空格进行分隔
    std::vector<std::string> tokens;
    std::istringstream tokenStream(str);
    std::string token;

    while (tokenStream >> token) {
        tokens.push_back(token);
    }

    return tokens;
}

std::string replace(const std::string& str, const std::string& oldStr, const std::string& newStr) {
    std::string result = str;
    size_t pos = 0;
    while ((pos = result.find(oldStr, pos)) != std::string::npos) {
        result.replace(pos, oldStr.length(), newStr);
        pos += newStr.length();
    }
    return result;
}

Scheduler::Scheduler(std::string tid_list_str, std::vector<int>& big_freq_list_in, std::vector<int>& little_freq_list_in) {
    this->tid_list_str = tid_list_str;
    std::string delimiter = ",";
    size_t pos = 0;
    while ((pos = tid_list_str.find(delimiter)) != std::string::npos) {
        std::string value = tid_list_str.substr(0, pos);
        tid_list.push_back(std::stoi(value));
        tid_list_str.erase(0, pos + delimiter.length());
    }
    tid_list.push_back(std::stoi(tid_list_str));
    thread_num = tid_list.size();

    this->big_freq_list.resize(pmu_num);
    this->little_freq_list.resize(pmu_num);

    this->big_freq_list = big_freq_list_in;
    this->little_freq_list = little_freq_list_in;

    all_pmu_data.resize(10);
    main_pmu_data.resize(10);
    sf_pmu_data.resize(10);
    render_pmu_data.resize(10);

    for (int i = 0; i < thread_num; i++) {
        set_cpu_mask_all(tid_list[i]);
    }

    for (int i = 0; i < thread_num; i++) {
        std::string command;
        if (position_bit[i] == 1) {
            set_cpu_mask_big(tid_list[i]);
        } else {
            set_cpu_mask_little(tid_list[i]);
        }
    }
}

// 定义性能事件类型
enum PerfEventType {
    TASK_CLOCK = PERF_TYPE_SOFTWARE,
    INSTRUCTIONS = PERF_TYPE_HARDWARE,
};

// 获取指定PID的性能事件计数器值
void getEventCounter(pid_t pid, PerfEventType eventType, long long& counter) {
    // 创建 perf_event_attr 结构体并设置属性
    struct perf_event_attr pe = {};
    std::memset(&pe, 0, sizeof(struct perf_event_attr));
    pe.type = eventType;
    pe.size = sizeof(struct perf_event_attr);
    pe.config = eventType == INSTRUCTIONS ? PERF_COUNT_HW_INSTRUCTIONS : PERF_COUNT_SW_TASK_CLOCK;
    pe.disabled = 1;
    pe.exclude_kernel = 1;
    pe.exclude_hv = 1;

    // 创建 perf_event_open 的文件描述符
    int fd = syscall(__NR_perf_event_open, &pe, pid, -1, -1, 0);
    if (fd == -1) {
        std::perror("perf_event_open");
        return;
    }

    // 启动事件计数
    ioctl(fd, PERF_EVENT_IOC_RESET, 0);
    ioctl(fd, PERF_EVENT_IOC_ENABLE, 0);

    // 停止事件计数
    sleep(1);

    // 读取事件计数器的值
    read(fd, &counter, sizeof(long long));

    // 关闭文件描述符
    close(fd);
}
 

std::vector<double> Scheduler::schdule(int big_current_freq, int little_current_freq) {
    big_scale = 0;
    little_scale = 0;
    big_scale_freq_idx = 0;
    little_scale_freq_idx = 0;

    std::vector<pid_t> pidList = tid_list;  // 要监测的PID列表

    std::vector<long long> taskClockList(pidList.size());
    std::vector<long long> instructionsList(pidList.size());

    std::vector<std::thread> threads;

    // 创建线程获取每个PID的计数器值
    for (size_t i = 0; i < pidList.size() * 2; ++i) {
        threads.emplace_back([&taskClockList, &instructionsList, i, pidList]() {
            getEventCounter(pidList[i], TASK_CLOCK, taskClockList[i]);
            // getEventCounter(pidList[i], INSTRUCTIONS, instructionsList[i]);
        });
        threads.emplace_back([&taskClockList, &instructionsList, i, pidList]() {
            // getEventCounter(pidList[i], TASK_CLOCK, taskClockList[i]);
            getEventCounter(pidList[i], INSTRUCTIONS, instructionsList[i]);
        });
    }

    // 等待所有线程完成
    for (auto& thread : threads) {
        thread.join();
    }

    // 打印结果
    for (size_t i = 0; i < pidList.size(); ++i) {
        all_pmu_data[i].push_back(taskClockList[i]/100000);
        all_pmu_data[i].push_back(instructionsList[i]);
    }


    for (int i1 = 0; i1 < thread_num; i1++) {
        priority_arr.push_back(all_pmu_data[i1][0] / 1000.0);
    }


    little_priority = 0;
    big_priority = 0;

    for (int i1 = 0; i1 < thread_num; i1++) {
        if (position_bit[i1] == 0) {
            little_priority += priority_arr[i1];
        } else {
            big_priority += priority_arr[i1];
        }
    }

    std::vector<double> a = priority_arr;
    std::vector<int> sort_index = argsort(a);
    // std::reverse(sort_index.begin(), sort_index.end());

    std::vector<int> big_scale_arr = {0, 0, 0};
    std::vector<int> little_scale_arr = {0, 0, 0};
    std::vector<int> big_scale_freq_arr = {0, 0, 0};
    std::vector<int> little_scale_freq_arr = {0, 0, 0};

    for (int i = 0; i < sort_index.size(); i++) {
        int index = sort_index[i];
        int task_clock = all_pmu_data[index][0];
        int inst = all_pmu_data[index][1];
        int big_scale_freq = 0;
        int little_scale_freq = 0;

        if (inst > inst_min[index]) {
            if (position_bit[index] == 1 && task_clock < task_clock_min_big[index]) {
                big_scale_freq = big_freq_list[big_freq_list.size() - 1];
                for (int j = big_freq_list.size() - 1; j >= 0; j--) {
                    int big_freq = big_freq_list[index];
                    if (big_freq / big_current_freq < task_clock / task_clock_min_big[index]) {
                        big_scale_freq = big_freq;
                        break;
                    }
                }
                if (big_scale_freq == big_freq_list[big_freq_list.size() - 1]) {
                    little_priority += priority_arr[index];
                    big_priority -= priority_arr[index];
                    position_bit[index] = 0;
                    set_cpu_mask_little(tid_list[index]);
                    break;
                } else {
                    big_scale_arr[i] = -1;
                    big_scale_freq_arr[i] = big_scale_freq;
                }
            } else if (position_bit[index] == 0 && task_clock < task_clock_min_little[index]) {
                little_scale_arr[i] = -1;
                little_scale_freq = little_freq_list[little_freq_list.size() - 1];
                for (int j = little_freq_list.size() - 1; j >= 0; j--) {
                    int little_freq = little_freq_list[j];
                    if (little_freq / little_current_freq < task_clock / task_clock_min_little[index]) {
                        little_scale_freq = little_freq;
                        break;
                    }
                }
                if (little_scale_freq == little_freq_list[little_freq_list.size() - 1]) {
                    little_scale_freq_arr[i] = little_freq_list[0];
                } else {
                    little_scale_freq_arr[i] = little_scale_freq;
                }
            } else if (position_bit[index] == 1 && task_clock > task_clock_max_big[index]) {
                big_scale_arr[i] = 1;
                big_scale_freq = big_freq_list[0];
                for (const int big_freq : big_freq_list) {
                    if (big_freq / big_current_freq > task_clock / task_clock_max_big[index]) {
                        big_scale_freq = big_freq;
                        break;
                    }
                }
                if (big_scale_freq == big_freq_list[0]) {
                    big_scale_freq_arr[i] = big_freq_list[big_freq_list.size() - 1];
                } else {
                    big_scale_freq_arr[i] = big_scale_freq;
                }
            } else if (position_bit[index] == 0 && task_clock < task_clock_min_little[index]) {
                little_scale_freq = little_freq_list[0];
                for (const int little_freq : little_freq_list) {
                    if (little_freq / little_current_freq > task_clock / task_clock_max_little[index]) {
                        little_scale_freq = little_freq;
                        break;
                    }
                }
                if (little_scale_freq == little_freq_list[0]) {
                    big_priority += priority_arr[index];
                    little_priority -= priority_arr[index];
                    position_bit[index] = 1;
                    set_cpu_mask_big(tid_list[index]);
                    break;
                } else {
                    little_scale_arr[i] = 1;
                    little_scale_freq_arr[i] = little_scale_freq;
                }
            }
        } else if (inst < inst_min[index]) {
            if (position_bit[index] == 0) {
                little_scale_freq = little_freq_list[0];
                for (const int little_freq : little_freq_list) {
                    if (little_freq / little_current_freq > inst_min[index] / inst) {
                        little_scale_freq = little_freq;
                        break;
                    }
                }
                if (little_scale_freq == little_freq_list[0]) {
                    big_priority += priority_arr[index];
                    little_priority -= priority_arr[index];
                    position_bit[index] = 1;
                    set_cpu_mask_big(tid_list[index]);
                    break;
                } else {
                    little_scale_arr[i] = 1;
                    little_scale_freq_arr[i] = little_scale_freq;
                }
            } else {
                big_scale_arr[i] = 1;
                big_scale_freq = big_freq_list[0];
                for (const int big_freq : big_freq_list) {
                    if (big_freq / big_current_freq > inst_min[index] / inst) {
                        big_scale_freq = big_freq;
                        break;
                    }
                }
                if (big_scale_freq == big_freq_list[0]) {
                    big_scale_freq_arr[i] = big_freq_list[big_freq_list.size() - 1];
                } else {
                    big_scale_freq_arr[i] = big_scale_freq;
                }
            }
        }
    }
    int down_max_freq = big_freq_list[0];
    int min_freq = big_freq_list[big_freq_list.size() - 1];
    int max_freq = big_freq_list[0];
    bool up_bool = false;
    bool down_bool = false;

    for (int i1 = 0; i1 < big_scale_arr.size(); i1++) {
        if (big_scale_arr[i1] == 1) {
            max_freq = std::max(max_freq, big_scale_freq_arr[i1]);
            up_bool = true;
        } else if (big_scale_arr[i1] == -1) {
            down_max_freq = std::max(down_max_freq, big_scale_freq_arr[i1]);
            down_bool = true;
        }
    }

    if (up_bool) {
        big_scale = 1;
        big_scale_freq = max_freq;
    } else if (down_bool) {
        big_scale = -1;
        big_scale_freq = down_max_freq;
    }

    down_max_freq = little_freq_list[0];
    min_freq = little_freq_list[little_freq_list.size() - 1];
    max_freq = little_freq_list[0];
    up_bool = false;
    down_bool = false;

    for (int i1 = 0; i1 < little_scale_arr.size(); i1++) {
        if (little_scale_arr[i1] == 1) {
            max_freq = std::max(max_freq, little_scale_freq_arr[i1]);
            up_bool = true;
        } else if (little_scale_arr[i1] == -1) {
            down_max_freq = std::max(down_max_freq, little_scale_freq_arr[i1]);
            down_bool = true;
        }
    }

    if (up_bool) {
        little_scale = 1;
        little_scale_freq = max_freq;
    }
    if (down_bool) {
        little_scale = -1;
        little_scale_freq = down_max_freq;
    }

    if (big_scale != 0) {
        for (int i1 = 0; i1 < big_freq_list.size(); i1++) {
            if (big_freq_list[i1] == big_scale_freq) {
                big_scale_freq_idx = i1;
            }
        }
    }
    if (little_scale != 0) {
        for (int i1 = 0; i1 < little_freq_list.size(); i1++) {
            if (little_freq_list[i1] == little_scale_freq) {
                little_scale_freq_idx = i1;
            }
        }
    }

    commu_info.push_back(big_priority);
    commu_info.push_back(little_priority);
    commu_info.push_back(big_scale);
    commu_info.push_back(big_scale_freq_idx);
    commu_info.push_back(little_scale);
    commu_info.push_back(little_scale_freq_idx);

    return commu_info;
}

File: application/app/src/main/java/com/example/networktrans/Scheduler.java
//package com.example.networktrans;
//
//import static java.lang.Math.max;
//import static java.lang.Math.round;
//
//import org.checkerframework.checker.units.qual.A;
//
//import java.util.ArrayList;
//import java.util.Arrays;
//import java.util.Collections;
//import java.util.Objects;
//import java.util.stream.Collectors;
//
//public class Scheduler {
//    private int interval = 5;
//    private int pmu_num = 2;
//    private int thread_num;
//    private ArrayList<Double> priority_arr = new ArrayList<>();
//    private ArrayList<ArrayList<Integer>> main_pmu_data = new ArrayList<>();
//    private ArrayList<ArrayList<Integer>> render_pmu_data = new ArrayList<>();
//    private ArrayList<ArrayList<Integer>> sf_pmu_data = new ArrayList<>();
//    private ArrayList<ArrayList<Integer>> all_pmu_data = new ArrayList<>();
//    private ArrayList<Double> commu_info = new ArrayList<>();
//
//    private int [] inst_min = {18700000, 155000000, 27700000};
//    private int [] task_clock_max_little = {122, 400, 131};
//    private int [] task_clock_max_big = {64, 400, 130};
//
//    private int [] task_clock_min_little = {94, 212, 104};
//    private int [] task_clock_min_big = {64, 467, 105};
//
//    private ArrayList<Integer> tid_list = new ArrayList<>();
//    private String tid_list_str;
//
//
//    private int big_scale = 0;
//    private int little_scale= 0;
//    private int big_scale_freq_idx = 0;
//    private int little_scale_freq_idx = 0;
//    private int big_scale_freq;
//    private int little_scale_freq;
//
//    private int [] big_freq_list;
//    private int [] little_freq_list;
//
//    private double big_priority = 0;
//    private double little_priority = 1;
//    private int [] position_bit = {1,1,0};
//
//    public Scheduler(String tid_list_str, int[] big_freq_list, int[] little_freq_list){
//        this.tid_list_str = tid_list_str;
//        for(final String value: tid_list_str.split(",")){
//            this.tid_list.add(Integer.parseInt(value));
//        }
//        thread_num = tid_list.size();
//
//        this.big_freq_list =  big_freq_list;
//        this.little_freq_list = little_freq_list;
//
//        initArrayList(all_pmu_data,10);
//        initArrayList(main_pmu_data,10);
//        initArrayList(sf_pmu_data,10);
//        initArrayList(render_pmu_data,10);
//
//
//        for(int i = 0; i < thread_num; i++){
//            String command = "taskset -ap ff " + tid_list.get(i);
//            CommandExecution.easyExec(command,true);
//        }
//
//        for(int i = 0; i < thread_num; i++){
//            String command;
//            if(position_bit[i] == 1){
//                command = "taskset -p f0 " + tid_list.get(i);
//            } else {
//                command = "taskset -p 0f " + tid_list.get(i);
//            }
//            CommandExecution.easyExec(command,true);
//        }
//    }
//    public  ArrayList<Double> schedule(int big_current_freq, int little_current_freq){
//        big_scale = 0;
//        little_scale = 0;
//        big_scale_freq_idx = 0;
//        little_scale_freq_idx = 0;
//        String command = "simpleperf stat -e task-clock,instructions -t " + tid_list_str + " --duration 1 --per-thread";
//        String result = CommandExecution.execCommand(command,true).successMsg;
//        String []  result_array = result.split("\n");
//        result_array = Arrays.copyOfRange(result_array,3,result_array.length);
//
//        if(result_array.length!=8){
//            commu_info.add(big_priority);
//            commu_info.add(little_priority);
//            commu_info.add((double)big_scale);
//            commu_info.add((double)big_scale_freq_idx);
//            commu_info.add((double)little_scale);
//            commu_info.add((double)little_scale_freq_idx);
//            return commu_info;
//        }
//        int pmu;
//        int i;
//        for(i = 0; i < pmu_num;i++) {
//            String thread_name = result_array[i * thread_num].split("\\s+")[1];
//            if (result_array[i * thread_num].contains("(ms)")) {
//                pmu = Integer.parseInt(result_array[i * thread_num].split("\\s+")[4].split("\\.")[0].replace(",", ""));
//            } else {
//                pmu = Integer.parseInt(result_array[i * thread_num].split("\\s+")[4].replace(",", ""));
//            }
//            if (Objects.equals(thread_name, "droid.ugc.aweme")) {
//                main_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(0).add(pmu);
//            } else if (Objects.equals(thread_name, "RenderThread")) {
//                render_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(1).add(pmu);
//            } else {
//                sf_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(2).add(pmu);
//            }
//
//            thread_name = result_array[i * thread_num + 1].split("\\s+")[1];
//            if (result_array[i * thread_num + 1].contains("(ms)")) {
//                pmu = Integer.parseInt(result_array[i * thread_num + 1].split("\\s+")[4].split("\\.")[0].replace(",", ""));
//            } else {
//                pmu = Integer.parseInt(result_array[i * thread_num + 1].split("\\s+")[4].replace(",", ""));
//            }
//            if (Objects.equals(thread_name, "droid.ugc.aweme")) {
//                main_pmu_data.get(i).add(pmu);   // TODO 不一定可以
//                all_pmu_data.get(0).add(pmu);
//            } else if (Objects.equals(thread_name, "RenderThread")) {
//                render_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(1).add(pmu);
//            } else {
//                sf_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(2).add(pmu);
//            }
//
//            thread_name = result_array[i * thread_num + 2].split("\\s+")[1];
//            if (result_array[i * thread_num + 2].contains("(ms)")) {
//                pmu = Integer.parseInt(result_array[i * thread_num + 2].split("\\s+")[4].split("\\.")[0].replace(",", ""));
//            } else {
//                pmu = Integer.parseInt(result_array[i * thread_num + 2].split("\\s+")[4].replace(",", ""));
//            }
//            if (Objects.equals(thread_name, "droid.ugc.aweme")) {
//                main_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(0).add(pmu);
//            } else if (Objects.equals(thread_name, "RenderThread")) {
//                render_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(1).add(pmu);
//            } else {
//                sf_pmu_data.get(i).add(pmu);
//                all_pmu_data.get(2).add(pmu);
//            }
//        }
//
//        for(int i1 = 0; i1 < thread_num; i1++){
//            priority_arr.add(all_pmu_data.get(i1).get(0) / 1000.0);
//        }
//
//        little_priority = 0;
//        big_priority = 0;
//
//        for(int i1 = 0; i1 < thread_num; i1++){
//            if(position_bit[i1] == 0){
//                little_priority = little_priority + priority_arr.get(i1);
//            }else{
//                big_priority = big_priority + priority_arr.get(i1);
//            }
//        }
//
//        ArrayList<Double> a = (ArrayList<Double>) priority_arr.clone();
//        ArrayList<Integer> sort_index = argsort(a);
//        Collections.reverse(sort_index);
//
//        int[] big_scale_arr = {0,0,0};
//        int[] little_scale_arr = {0,0,0};
//        int[] big_scale_freq_arr = {0,0,0};
//        int[] little_scale_freq_arr = {0,0,0};
//
//        for(i = 0; i < sort_index.size(); i++){
//            int index = sort_index.get(i);
//            int task_clock = all_pmu_data.get(index).get(0);
//            int inst = all_pmu_data.get(index).get(1);
//            int big_scale_freq = 0;
//            int little_scale_freq= 0;
//
//            if(inst > inst_min[index]){
//                if(position_bit[index] == 1 && task_clock < task_clock_min_big[index]){
//                    big_scale_freq = big_freq_list[big_freq_list.length-1];
//                    for(int j = big_freq_list.length -1; j >=0; j--){
//                        int big_freq = big_freq_list[index];
//                        if(big_freq / big_current_freq < task_clock / task_clock_min_big[index]){
//                            big_scale_freq = big_freq;
//                            break;
//                        }
//                    }
//                    if(big_scale_freq == big_freq_list[big_freq_list.length - 1]){
//                       little_priority = little_priority + priority_arr.get(index);
//                       big_priority = big_priority - priority_arr.get(index);
//                       position_bit[index] = 0;
//                       command = "taskset -p 0f "+Integer.toString(tid_list.get(index));
//                       CommandExecution.easyExec(command,true);
//                       break;
//                    }else {
//                        big_scale_arr[i] = -1;
//                        big_scale_freq_arr[i] = big_scale_freq;
//                    }
//
//                } else if(0 == position_bit[index] && task_clock < task_clock_min_little[index]){ //250
//                    little_scale_arr[i] = -1;
//                    little_scale_freq = little_freq_list[little_freq_list.length - 1];
//                    for(int j = little_freq_list.length-1; j >= 0; j--){
//                        int little_freq = little_freq_list[j];
//                        if(little_freq / little_current_freq < task_clock / task_clock_min_little[index]){
//                            little_scale_freq = little_freq;
//                            break;
//                        }
//                    }
//                    if(little_scale_freq == little_freq_list[little_freq_list.length - 1]){
//                        little_scale_freq_arr[i] = little_freq_list[0];
//                    }else{
//                        little_scale_freq_arr[i] = little_scale_freq;
//                    }
//                } else if(position_bit[index] == 1 && task_clock > task_clock_max_big[index]){
//                    big_scale_arr[i] = 1;
//                    big_scale_freq = big_freq_list[0];
//                    for(final  int big_freq : big_freq_list){
//                        if(big_freq / big_current_freq > task_clock / task_clock_max_big[index]){
//                            big_scale_freq = big_freq;
//                            break;
//                        }
//                    }
//                    if (big_scale_freq == big_freq_list[0]){
//                        big_scale_freq_arr[i] = big_freq_list[big_freq_list.length - 1];
//                    } else {
//                        big_scale_freq_arr[i] = big_scale_freq;
//                    }
//                } else if(position_bit[index] == 0 && task_clock < task_clock_min_little[index]){
//                    little_scale_freq = little_freq_list[0];
//                    for(final int little_freq:little_freq_list){
//                        if(little_freq / little_current_freq > task_clock / task_clock_max_little[index]){
//                            little_scale_freq = little_freq;
//                            break;
//                        }
//                    }
//                    if(little_scale_freq == little_freq_list[0]){
//                        big_priority = big_priority + priority_arr.get(index);
//                        little_priority = little_priority - priority_arr.get(index);
//                        position_bit[index]= 1;
//                        command = "taskset -p f0" + tid_list.get(index);
//                        result = CommandExecution.execCommand(command,true).successMsg;
//                        break;
//                    } else {
//                        little_scale_arr[i] = 1;
//                        little_scale_freq_arr[i] = little_scale_freq;
//                    }
//                }
//            } else if(inst < inst_min[index]){  // 310
//                if(position_bit[index] == 0){
//                    little_scale_freq = little_freq_list[0];
//                    for(final int little_freq:little_freq_list){
//                        if(little_freq / little_current_freq > inst_min[index] / inst){
//                            little_scale_freq = little_freq;
//                            break;
//                        }
//                    }
//                    if(little_scale_freq == little_freq_list[0]){
//                        big_priority = big_priority + priority_arr.get(index);
//                        little_priority = little_priority - priority_arr.get(index);
//                        position_bit[index] = 1;
//                        command = "taskset -p f0 " + Integer.toString(tid_list.get(index));
//                        CommandExecution.execCommand(command,true);
//                        break;
//                    } else {
//                        little_scale_arr[i] = 1;
//                        little_scale_freq_arr[i] = little_scale_freq;
//                    }
//                }else{
//                    big_scale_arr[i] = 1;
//                    big_scale_freq = big_freq_list[0];
//                    for(final int big_freq: big_freq_list){
//                        if(big_freq / big_current_freq > inst_min[index] / inst){
//                            big_scale_freq = big_freq;
//                            break;
//                        }
//                    }
//                    if(big_scale_freq == big_freq_list[0]){
//                        big_scale_freq_arr[i] = big_freq_list[big_freq_list.length - 1];
//                    }else {
//                        big_scale_freq_arr[i] = big_scale_freq;
//                    }
//                }
//            }
//        }
//
//        int down_max_freq = big_freq_list[0];
//        int min_freq = big_freq_list[big_freq_list.length-1];
//        int max_freq = big_freq_list[0];
//        boolean up_bool = false;
//        boolean down_bool = false;
//        for(int i1 = 0; i1 < big_scale_arr.length; i1++){
//            if(big_scale_arr[i1] == 1){
//                max_freq = max(max_freq, big_scale_freq_arr[i1]);
//                up_bool = true;
//            }else if(big_scale_arr[i1] == -1){
//                down_max_freq = max(down_max_freq, big_scale_freq_arr[i1]);
//                down_bool = true;
//            }
//        }
//
//        if(up_bool){
//            big_scale = 1;
//            big_scale_freq = max_freq;
//        }else if(down_bool){
//            big_scale=  -1;
//            big_scale_freq = down_max_freq;
//        }
//
//        down_max_freq = little_freq_list[0];
//        min_freq = little_freq_list[little_freq_list.length - 1];
//        max_freq = little_freq_list[0];
//        up_bool = false;
//        down_bool = false;
//
//
//        for(int i1 = 0 ; i1 < little_scale_arr.length; i1++){
//            if(little_scale_arr[i1] == 1){
//                max_freq = max(max_freq, little_scale_freq_arr[i1]);
//                up_bool = true;
//            } else if(little_scale_arr[i1] == -1){
//                down_max_freq = max(down_max_freq, little_scale_freq_arr[i1]);
//                down_bool = true;
//            }
//        }
//        if(up_bool){
//            little_scale = 1;
//            little_scale_freq = max_freq;
//        }
//        if(down_bool){
//            little_scale = -1;
//            little_scale_freq = down_max_freq;
//        }
//
//        if(big_scale != 0){
//            for(int i1 = 0; i1 < big_freq_list.length; i1++){
//                if(big_freq_list[i1] == big_scale_freq){
//                    big_scale_freq_idx = i1;
//                }
//            }
//        }
//        if(little_scale != 0){
//            for(int i1 =0 ; i1< little_freq_list.length; i1++){
//                if(little_freq_list[i1] == little_scale_freq){
//                    little_scale_freq_idx = i1;
//                }
//            }
//        }
//
//        commu_info.add(big_priority);
//        commu_info.add(little_priority);
//        commu_info.add((double)big_scale);
//        commu_info.add((double)big_scale_freq_idx);
//        commu_info.add((double)little_scale);
//        commu_info.add((double)little_scale_freq_idx);
//
//        return commu_info;
//    }
//
//    private ArrayList<Integer> argsort(ArrayList<Double> input) {
//            int h = input.size();
//            ArrayList<Integer> index = new ArrayList<>();
//
//            for (int i = 0; i < h; ++i)
//                index.add(i);
//
//            for (int i = h - 1; i > 0; --i)
//            {
//                int min = 0;
//                for (int j = 1; j <= i; ++j)
//                    if (input.get(index.get(j)) < input.get(index.get(min)))
//                        min = j;
//
//                int temp = index.get(i);
//                index.set(i, index.get(min));
//                index.set(min, temp);
//            }
//            return index;
//    }
//
//    private void initArrayList(ArrayList<ArrayList<Integer>> a, int num){
//        for(int i  = 0; i < num; i++){
//            ArrayList<Integer> temp = new ArrayList<>();
//            a.add(temp);
//        }
//    }
//}
